agent/consul/state/catalog.go

```diff
@@ -4535,31 +4535,46 @@ func (s *Store) ServiceTopology(
 		maxIdx = idx
 	}
 
-	// Store downstreams with at least one instance in transparent proxy mode.
+	// Store downstreams with at least one instance in transparent proxy or connect native mode.
 	// This is to avoid returning downstreams from intentions when none of the downstreams are transparent proxies.
-	tproxyMap := make(map[structs.ServiceName]struct{})
+	proxyMap := make(map[structs.ServiceName]struct{})
 	for _, downstream := range unfilteredDownstreams {
 		if downstream.Service.Proxy.Mode == structs.ProxyModeTransparent {
 			sn := structs.NewServiceName(downstream.Service.Proxy.DestinationServiceName, &downstream.Service.EnterpriseMeta)
-			tproxyMap[sn] = struct{}{}
+			proxyMap[sn] = struct{}{}
+		}
+		if downstream.Service.Connect.Native {
+			sn := downstream.Service.CompoundServiceName()
+			proxyMap[sn] = struct{}{}
 		}
 	}
 
 	var downstreams structs.CheckServiceNodes
+	refsToRemove := make(map[structs.ServiceName]struct{})
+	servicesSeen := make(map[string]struct{})
 	for _, downstream := range unfilteredDownstreams {
+		serviceNodeID := fmt.Sprintf("%s/%s", downstream.Node.Node, downstream.Service.ID)
```

Comment:
@kisunji Is there a better way to dedupe services than by `Node.Node` & `Service.ID`. Interestingly enough `Node.ID` is empty, is that due to the tests or is that expected to be empty?

Reply:
`Node.ID` is optional so checking `Node.Node` & `Service.ID` should be sufficient for uniqueness

Reply:
I responded to a separate comment but could this deduping happen within `combinedServiceNodesTxn`?

The implementation should be simpler in that function:
* Instead of appending `csn` to `resp` directly, add to a `map[{node+serviceid}]structs.CheckServiceNode`
* After finishing the loop over `names`, loop through the map and add to `resp`